### ISSUE TITLE
gcc 5.2.0 segfaults when compiling `#include <cpp-sort/sorters.h>`

### ISSUE THREAD

**Issue by FelipeLema** [Morwenn/cpp-sort#162]
How to reproduce (paths were altered for security concerns):
```
# echo '#include <cpp-sort/sorters.h>' > single.cpp
# g++ -I/path/to/cpp-sort/1.6.0/include -std=gnu++14 single.cpp -o single.o

In file included from /path/to/cpp-sort/1.6.0/include/cpp-sort/sorters/default_sorter.h:32:0,
                 from /path/to/cpp-sort/1.6.0/include/cpp-sort/sorters.h:32,
                 from single.cpp:1:
/path/to/cpp-sort/1.6.0/include/cpp-sort/adapters/self_sort_adapter.h:159:21: internal compiler error: Segmentation fault
             detail::has_sort_method<Args...>,
                     ^
0xaa09af crash_signal
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/toplev.c:383
0x642565 dependent_template_arg_p
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/pt.c:21769
0x6426ae check_instantiated_arg
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/pt.c:15884
0x6428fc check_instantiated_args
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/pt.c:15958
0x6586a5 instantiate_template_1
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/pt.c:16056
0x6586a5 instantiate_template(tree_node*, tree_node*, int)
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/pt.c:16151
0x6e28e7 finish_id_expression(tree_node*, tree_node*, tree_node*, cp_id_kind*, bool, bool, bool*, bool, bool, bool, bool, char const**, unsigned int)
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/semantics.c:3581
0x695be4 cp_parser_primary_expression
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:4763
0x69809d cp_parser_postfix_expression
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:6157
0x69a389 cp_parser_unary_expression
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:7438
0x69aec7 cp_parser_binary_expression
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:8172
0x69b4cd cp_parser_assignment_expression
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:8430
0x69b8a3 cp_parser_constant_expression
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:8676
0x6a1b3e cp_parser_template_argument
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:14420
0x6a1b3e cp_parser_template_argument_list
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:14138
0x6a1b3e cp_parser_enclosed_template_argument_list
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:24070
0x6a2756 cp_parser_template_id
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:13821
0x6a2e56 cp_parser_class_name
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:19735
0x699036 cp_parser_qualifying_entity
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:5713
0x699036 cp_parser_nested_name_specifier_opt
	/path/to/gcc/sources/src/gcc-5.2.0/gcc/cp/parser.c:5398
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <http://gcc.gnu.org/bugs.html> for instructions.
# 
```

**Comment by FelipeLema** [comment]
BTW, this happened with 1.6.0

**Comment by Morwenn** [comment]
I guess that the CI is using a more recent version of GCC where the ICE was fixed. I will have a look at it later - if you can do you think that you could try with a more recent version of GCC 5? I would expect people to use the latest patch release.

**Comment by FelipeLema** [comment]
I cannot change my gcc version (provided by corporate).

I don't really think there's much to be done here. May I suggest that the README says "gcc 5.5" (latest release) instead of "gcc 5".

I'd say that's about enough for me.

**Comment by Morwenn** [comment]
That was my "quick fix" plan, I don't think I'm willing to maintain the template hell on non-latest-patch releases, especially since the latest plans for the library involve C++17 and C++20. I will still have a look and check whether it's possible to find a simple solution, but I don't have high hopes.

**Comment by Morwenn** [comment]
Commit 7bde4234d255f17a0f55dfa9dcf0f22159693395 mentions that only 5.5 support or later is guaranteed (it was supposed to mention this issue but the issue number was at the start of the comment line and was considered to be a comment...).

**Comment by FelipeLema** [comment]
thank you for considering this... closing it up!

**Comment by Morwenn** [comment]
I will still try to look whether there is a simple enough fix for 5.2 when I make time for it.
